### ISSUE TITLE
issue while running train.py

### ISSUE THREAD

**Issue by rohitsaluja22** [MichalBusta/DeepTextSpotter#17]
Hi we tried train.py but are getting following error:

>     I0213 04:00:44.575316 17258 net.cpp:242] This network produces output softmax
>     I0213 04:00:44.575346 17258 net.cpp:255] Network initialization done.
>     I0213 04:00:44.575435 17258 solver.cpp:56] Solver scaffolding done.
>     [416, 416]
>     I0213 04:00:44.581261 17258 solver.cpp:330] Iteration 0, Testing net (#0)
>     Floating point exception (core dumped)

We are creating a separate .txt file for each .jpeg file. For example, one of the .txt file looks like following:-

>     0 86.544325 113.70325 61.5000988376 10.8000627304 3.03217953614 calliberation

Our image size is 426 x 240.. shall we change it to some standard size?

Please help us resolving the error.

**Comment by rohitsaluja22** [comment]
Hey, we also tried resizing images to 240 x 240( though it will kill the aspect ratios which might not be recommended as per your paper). Still, we are getting the same error.

**Comment by leetwito** [comment]
Your labels aren't in the right format.
Look at this issue :
https://github.com/MichalBusta/DeepTextSpotter/issues/10
You haven't normalized the labels

**Comment by rohitsaluja22** [comment]
hey leetwito , thanks for your reply, I got it, I was not normalizing the coordinates and another mistake, the angle calculation was using the right-handed coordinate system, whereas the image follows the left-handed. That's why my angle was 3 radians.
Also, it's not mentioned here but I am giving image path as the first word in the line of train_list.txt. Hope I am right in that?

**Comment by rohitsaluja22** [comment]
Hey, I am getting the errors like following: 

> train/172884.jpeg
> bad gt line:  E^TmK^BGfO«+^@}

The actual line in train/172884.txt is:

> 0 0.504146713615 0.697329166667 0.178035531192 0.0295033225651 0.177802086545 TABLE-TENNIS

Please help

**Comment by MichalBusta** [comment]
Hi, 
bug in ondisk_data_layer:
...
 std::string gt_file = boost::replace_all_copy(file_name, ".jpg", ".txt");
 gt_file = boost::replace_all_copy(gt_file, ".png", ".txt");

the data reader accepts only .jpg and .png images.

**Comment by rohitsaluja22** [comment]
Hey, MichalBusta, thanks.. training is on now...  Also, it would be helpful if you can guide how to load the trained model in demo.py.

Also can I retrain on the model you have shared?

Thanks...

**Comment by leetwito** [comment]
on 'models.py' file, under function 'create_XXXX_solver' function, add the following line:
 solver.net.copy_from('PATH/TO/CAFFEMODEL/FILE/FILE_NAME.caffemodel')

**Comment by ayushbits** [comment]
Thanks @leetwito for the quick reply. I have copied as you said in models.py 
```
def create_tiny_yolo_solver():
  solver = caffe.get_solver('models/tiny_solver.prototxt')
  solver.net.copy_from('/home/ayush/DeepTextSpotter/backup/yolo_mobile_iter_25000.caffemodel')
  return solver
def create_recognizer_solver():
  solver = caffe.get_solver('models/solver_ctc.prototxt')
  #solver.restore('backup/recog_iter_195000.solverstate')
  solver.net.copy_from('/home/ayush/DeepTextSpotter/backup/yolo_mobile_iter_25000.caffemodel')
  return solver
def create_solvers_tiny():
  proposal_net = create_tiny_yolo_solver()
  recog = create_recognizer_solver()
  return proposal_net, recog
````
However, it is ignoring all source layers 
```
I0216 05:31:41.686627  4369 net.cpp:744] Ignoring source layer conv20/1x1/bn
I0216 05:31:41.686632  4369 net.cpp:744] Ignoring source layer conv20/1x1/scale
I0216 05:31:41.686636  4369 net.cpp:744] Ignoring source layer conv20/relu2
I0216 05:31:41.686640  4369 net.cpp:744] Ignoring source layer conv_22
I0216 05:31:41.686645  4369 net.cpp:744] Ignoring source layer RegionLayer
I0216 05:31:41.686648  4369 net.cpp:744] Ignoring source layer loss_RegionLayer_0_split
I0216 05:31:41.686652  4369 net.cpp:744] Ignoring source layer dummy
```
I have trained  the model until 500000 iterations. Is the model not trained on sufficient iterations? What could be the reason of not forming correct model after training. Can you point me to the reasons?

Thanks in Advance.

**Comment by MichalBusta** [comment]
Hello @rohitsaluja22 ,
the model is created in models.py  with: 
proposal_net = caffe.Net('{0}/models/tiny.prototxt'.format(baseDir), '{0}/models/tiny.caffemodel'.format(baseDir), phase)

so you need just change tiny.caffemodel for your model.

**Comment by rohitsaluja22** [comment]
Hey @MichalBusta, thanks for your answers.
We can see following files getting saved in "backup" folder:

> retrained_iter_XXXX.solverstate  retrained_iter_XXXX.caffemodel

retrained_iter_XXXX.caffemodel is for detection. Is retrained_iter_XXXX.solverstate the file for recognition? I guess not... if no, how to train for recognition simultaneously?

Thanks in advance
Rohit

**Comment by leetwito** [comment]
XXX.caffemodel is a weights file.
XXX.solverstate is a file that saves the current state of the solver (if the lr was changed etc)

you need only the .caffemodel files for your purposes. yolo is the detection net, and you should have also another caffemodel file for the OCR net. If you don't have another file saved to 'backup' folder, you should look in the solver_ctc.prototxt what's the snapshot rate is. If snapshot equals 1000, it means the system will export the solverstate and the caffemodel files every 1000 iterations. 
Did you try to download and use the weights posted in the 'Readme' file?
You should insert your caffemodel files paths in the 'create_XXX_solver' functions in 'models.py'

**Comment by ayushbits** [comment]
@leetwito Thanks for the response. 
My snapshot rates in both tiny_solver and solver_ctc is 4000 . Both are saved at backup/retrained and backup/recog respectively. Though, backup/retrained* is generated, backup/recog* is not generated

tiny_solver (excerpt)
```
max_iter: 10000
# snapshot intermediate results
snapshot: 4000
snapshot_prefix: "backup/retrained"
# solver mode: CPU or GPU
solver_mode: GPU
clip_gradients: 200
```
solver_ctc (excerpt)
```
display: 100
# The maximum number of iterations
max_iter: 4000
# snapshot intermediate results
snapshot: 4000
snapshot_prefix: "backup/recog"
# solver mode: CPU or GPU
solver_mode: GPU

```

**Comment by rohitsaluja22** [comment]
ayush, is it giving recog* after 4000 iterations?

**Comment by ayushbits** [comment]
Hi @MichalBusta @leetwito ,
Our training image size is 426x240, whereas train.py mention different image sizes. Do we need to change image size in train.py to 426x240 or system resizes it to proper training input size ?

While retraining, recognition model is not saved at every 4000th iteration despite snapshot: 4000. However, detector is saved at every snapshot iteration. Can you tell how to fix this ?

**Comment by leetwito** [comment]
hey @ayushbits 
did you try the solution @rohitsaluja22 offered? raise the max_iter to 10k or anything higher than your snapshop rate.
your input can be at any size, it will be resized by the system.

**Comment by rohitsaluja22** [comment]
Hey @MichalBusta @leetwito, we tried retraining over the model you have provided but are not getting better accuracy. Shall we retrain from scratch?
Also, will reducing vocabulary in variable "codec" in train.py while retraining help?
Also where do you generally stop while retraining, max E2E recall on validation data!

**Comment by MichalBusta** [comment]
On 02/28/2018 11:10 AM, rohitsaluja22 wrote:
>
> Hey @MichalBusta <https://github.com/michalbusta> @leetwito 
> <https://github.com/leetwito>, we tried retraining over the model you 
> have provided but are not getting better accuracy. Shall we retrain 
> from scratch?
>
depends - if you are using same training data as we were, there is no 
point - the model will have almost same accuracy.
>
> Also, will reducing vocabulary in variable "codec" in train.py while 
> retraining help?
>
you can try, but I don't thing so.
If you want better accuracy, you can do:
     - provide better (and more) training data
     - use better loss functions
     - use better back-bone network
>
> Also where do you generally stop while retraining, max E2E recall on 
> validation data
>
- 3 days of training - keep log, and pick the best.

**Comment by rohitsaluja22** [comment]
@MichalBusta thanks... 3 days of training! Are you talking about training from scratch? Please confirm. My question was about retraining after you train the detector and recognizer on synthetic data!

**Comment by whulc** [comment]
name: "1.0 MobileNet-128"
layer {
  name: "data"
  type: "OnDiskData"
  top: "data"
  input_param { shape: { dim: 1 dim: 1 dim: 32 dim: 32 } }
  data_param { source: "/home/busta/data/SynthText/train_icdar.txt" }
  top: "gt_boxes"
  top: "gt_labels"
  top: "line_boxes"
  top: "line_labels"
}
how should I organize the source file?

**Comment by 19931991** [comment]
How to get the data of training ? I don not understand the code of data.py , Thank you very much!@rohitsaluja22
